enhancements/hybrid-helm-operator.md

```diff
@@ -0,0 +1,223 @@
+  dependencies, automatic upgrades of helm versions, or any other conveniences provided
+  by the base image. The user will need to manage and upgrade their code by hand.
+
+## Proposal
+
+There are 3 stages to this proposal:
+
+1. Joe Lanford has made significant progress on this already, in the
+   [joelanford/helm-operator][helm-operator-repo] repository.
+   First we will migrate this code to the operator-sdk repo, where it will replace the
+   existing helm-operator code.
+1. We will write a new SDK plugin that can scaffold out a hybrid operator. As a first
+   iteration we can just scaffold out most of the code that is currently in the
+   [helm-operator][helm-operator-repo] (though the exact structure of the code will likely change)
+   and use that to generate what we ship in the base image.
+1. We will break as much code as we reasonably can out of the scaffolding and into a
+   utility library, to minimize the amount of code the user will need to maintain. This
+   library could either be the existing operator-lib package, or a new, helm-operator specific
+   package.
+
+### User Stories
+
+#### Story 1
+
+As a Helm-based operator author, I want to add additional programmatic backup and restore
+functionality to my operator, without starting over with a completely different type of operator.
+
+#### Story 2
+
+As a Helm-based operator author, I want to change core behavior of the reconciliation loop
+that is not currently exposed in the watches.yaml
+
+#### Story 3
+
+As a prospective author of a Helm-based operator, I want to know that if I invest time and
+effort creating a Helm-based operator, I won't need to start over in order to create a higher
+level operator.
+
+#### Story 4
+
+As a Golang-based operator author, I would like to leverage existing helm charts for deploying
+pieces of my application, without sacrificing the flexibility of Go.
+
+
+### Implementation Details/Notes/Constraints
+
+#### Scaffolding plugin for the SDK
+
+The basic idea behind the hybrid Helm/Go operator plugin is that we will scaffold out a Go
+operator project structure, with all of the code bits from the helm-operator project filled in,
+such that building the project without adding any APIs or modifying any code will result in the
+same container as what is shipped in the helm-operator base image.
+
+Because the driving need behind this functionality is the ability to customize any piece of
+the Helm-based operator, we will scaffold out every bit of code that is not accessible via
+a library, and err on the side of scaffolding more than we need to.
+
+For the first iteration of the hybrid Helm/Go operator plugin, we will likely just scaffold
+out nearly everything that is currently in the [helm-operator repository][helm-operator-repo], but
+with a `main.go` that is more in line with a typical Go-based operator, including the kubebuilder
+scaffolding markers that make it easy to add pure Go-based apis. This will provide maximal
```

Comment:
```
including the kubebuilder scaffolding markers that make it easy to add pure Go-based apis. 
```
Hmm.. do you mean the types.go for the chart? If so, please ignore my previous comment.
This seemed confusing because kb markers would be on the particular type to generate crd. So initial scaffold would be just `main.go` as in go operator.

Reply:
So it doesn't look like generating a types.go for the chart is going to be very easy, at least in the short-term (adoption of values.schema.yaml may change that), so this is talking just about APIs implemented in Go, not ones that make use of helm charts.

> initial scaffold would be just main.go as in go operator.

I think the initial scaffold would not be the same `main.go` that Go operators get, it would be one that has all the markers for adding Go apis but also calls the core logic for handling loading the `watches.yaml` and setting up watches/reconciliation based on that. Then adding additional Helm-backed APIs would only involve modifying the `watches.yaml` and `config/`, same as in a pure Helm operator.